### ISSUE TITLE
Slow when using undo

### ISSUE THREAD

**Issue by dariusrosendahl** [dsandstrom/atom-double-tag#6]
First off: Can't believe I finally found this package. Been missing this feature for a long time in editors except for notepad++/visual studio. _happy dance_

The only problem I'm having is that using CTRL+Z is really slow after I changed a tag, then it doesn't save time anymore.

**Comment by dsandstrom** [comment]
What do you mean by ctrl+z, undo?  Also, please clarify what you mean my "doesn't save time anymore".

**Comment by DoorDarius** [comment]
Yes, when I try to undo an action done by double that it's incredibly slow per character. Perhaps it would be a good idea to improve performance by changing the whole tag back with undo in 1 go instead of per character. 

Ah sorry, I just ment that it was so slow that it didn't save me the time your package was supposed to save me with this bug.

**Comment by dariusrosendahl** [comment]
Even with the fix it's unfortunately still slow. It works fine for example when you select a div and then make it into a span.

But when you delete the div (letter-by-letter) and then use undo it's really slow, like it's thinking after each letter.

**Comment by Sam-Hoult** [comment]
Same issue.
When using undo, it removes letter by letter on each opening and closing tag.
So to undo say editing `<a></a>` to `<button></button>`, takes hitting undo 17 - 19 times (it varies as sometime it won't even remove a letter)
Perhaps between starting and finishing should be a buffer that can be changed on undo. I'm not sure how atom handles items going into history but will have a look.

**Comment by dsandstrom** [comment]
I'll work on this issue next.

**Comment by JamesCoyle** [comment]
It should only save the state to the buffer in certain states, after a second or so of inactivity, or if the focus shifts away. If I just update a tag, undo and redo should toggle between the before and after states. If the user deletes the full tag name and retypes a new one it should save the old tag state, the empty tag, and the new tag state.

**Comment by mrleblanc101** [comment]
Is this going to be fixed one day ? This package is essential but this is so annoying.

**Comment by dsandstrom** [comment]
I've looked into the problem, but haven't found an elegant solution.  The basic problem is undo causes the package to see that tag changed and tries to mimic the changes in the sibling tag.  I need to ignore the changes that undo/redo makes.

The problem is when text changes with the tag, there isn't a way to know whether it was caused by regular input or by undo.  This would have to added to Atom's API.

**Comment by tujoworker** [comment]
Have a look at this solution: https://atom.io/packages/js-refactor

**Comment by dsandstrom** [comment]
Thanks, looks like the real work is done in https://github.com/hax/refactor .

https://github.com/hax/refactor/blob/c42048e5b31215e22b3b6f5563fe3dd28f70f66f/lib/watcher.coffee#L185
```coffee
@editor.setTextInBufferRange [start, start], marker.delimiter, undo: 'skip'
```

Hmm, something like that might work.  Have the text that this package adds to the sibling tag be ignored by undo.  That way when someone hits undo, Atom just undos the tag the cursor was on and this package copies the new text to the sibling tag.

**Comment by dsandstrom** [comment]
Fixed by v1.0.0
